# Importing Data

In [10]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [11]:
df = pd.read_csv("Movie_regression.csv", header = 0)

In [19]:
df

Marketing expense  Production expense  Multiplex coverage     Budget  \
0              20.1264               59.62               0.462  36524.125   
1              20.5462               69.14               0.531  35668.655   
2              20.5458               69.14               0.531  39912.675   
3              20.6474               59.36               0.542  38873.890   
4              21.3810               59.36               0.542  39701.585   
..                 ...                 ...                 ...        ...   
501            21.2526               78.86               0.427  36624.115   
502            20.9054               78.86               0.427  33996.600   
503            21.2152               78.86               0.427  38751.680   
504            22.1918               78.86               0.427  37740.670   
505            20.9482               78.86               0.427  33496.650   

     Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0           138.7               7.825                8.095            7.910   
1           152.4               7.505                7.650            7.440   
2           134.6               7.485                7.570            7.495   
3           119.3               6.895                7.035            6.920   
4           127.7               6.920                7.070            6.815   
..            ...                 ...                  ...              ...   
501         142.6               8.680                8.775            8.620   
502         150.2               8.780                8.945            8.770   
503         164.5               8.830                8.970            8.855   
504         162.8               8.730                8.845            8.800   
505         154.3               8.640                8.880            8.680   

     Producer_rating  Critic_rating  Trailer_views 3D_available  Time_taken  \
0              7.995           7.94         527367          YES      109.60   
1              7.470           7.44         494055           NO      146.64   
2              7.515           7.44         547051           NO      147.88   
3              7.020           8.26         516279          YES      185.36   
4              7.070           8.26         531448           NO      176.48   
..               ...            ...            ...          ...         ...   
501            8.970           6.80         492480           NO      186.96   
502            8.930           7.80         482875          YES      132.24   
503            9.010           7.80         532239           NO      109.56   
504            8.845           6.80         496077          YES      158.80   
505            8.790           6.80         518438          YES      205.60   

     Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0            223.840  Thriller              23            494       48000  
1            243.456     Drama              42            462       43200  
2           2022.400    Comedy              38            458       69400  
3            225.344     Drama              45            472       66800  
4            225.792     Drama              55            395       72400  
..               ...       ...             ...            ...         ...  
501          243.584    Action              27            561       44800  
502          263.296    Action              20            600       41200  
503          243.824    Comedy              31            576       47800  
504          303.520    Comedy              47            607       44000  
505          203.040    Comedy              45            604       38000  

[506 rows x 18 columns]

In [20]:
df.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0         138.7               7.825                8.095            7.910   
1         152.4               7.505                7.650            7.440   
2         134.6               7.485                7.570            7.495   
3         119.3               6.895                7.035            6.920   
4         127.7               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views 3D_available  Time_taken  \
0            7.995           7.94         527367          YES      109.60   
1            7.470           7.44         494055           NO      146.64   
2            7.515           7.44         547051           NO      147.88   
3            7.020           8.26         516279          YES      185.36   
4            7.070           8.26         531448           NO      176.48   

   Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0          223.840  Thriller              23            494       48000  
1          243.456     Drama              42            462       43200  
2         2022.400    Comedy              38            458       69400  
3          225.344     Drama              45            472       66800  
4          225.792     Drama              55            395       72400

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Marketing expense    506 non-null    float64
 1   Production expense   506 non-null    float64
 2   Multiplex coverage   506 non-null    float64
 3   Budget               506 non-null    float64
 4   Movie_length         506 non-null    float64
 5   Lead_ Actor_Rating   506 non-null    float64
 6   Lead_Actress_rating  506 non-null    float64
 7   Director_rating      506 non-null    float64
 8   Producer_rating      506 non-null    float64
 9   Critic_rating        506 non-null    float64
 10  Trailer_views        506 non-null    int64  
 11  3D_available         506 non-null    object 
 12  Time_taken           506 non-null    float64
 13  Twitter_hastags      506 non-null    float64
 14  Genre                506 non-null    object 
 15  Avg_age_actors       506 non-null    int

we have 494 values in 'Time_taken' variable, so we'll have to fill the missing values

we can fill the missing values with the mean of the 'Time_taken' variable just for the ease of calculation

# Treating missing values

In [22]:
df['Time_taken'].mean()  #first we calculate the mean

157.39149797570852

In [23]:
df['Time_taken'].fillna(value = df['Time_taken'].mean(), inplace = True) #we use fillna to fill missing values, value will take mean and inplace will permanently replace the blank values with our mean values in the dataframe

In [24]:
df.info()  #now we check and see that the 'Time_taken' variable has 506 non-null values and the missing values were filled

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Marketing expense    506 non-null    float64
 1   Production expense   506 non-null    float64
 2   Multiplex coverage   506 non-null    float64
 3   Budget               506 non-null    float64
 4   Movie_length         506 non-null    float64
 5   Lead_ Actor_Rating   506 non-null    float64
 6   Lead_Actress_rating  506 non-null    float64
 7   Director_rating      506 non-null    float64
 8   Producer_rating      506 non-null    float64
 9   Critic_rating        506 non-null    float64
 10  Trailer_views        506 non-null    int64  
 11  3D_available         506 non-null    object 
 12  Time_taken           506 non-null    float64
 13  Twitter_hastags      506 non-null    float64
 14  Genre                506 non-null    object 
 15  Avg_age_actors       506 non-null    int

# Creating dummy variable

Now we want to create some dummy variables to convert our categorical data into numerical data

In [25]:
df.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0         138.7               7.825                8.095            7.910   
1         152.4               7.505                7.650            7.440   
2         134.6               7.485                7.570            7.495   
3         119.3               6.895                7.035            6.920   
4         127.7               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views 3D_available  Time_taken  \
0            7.995           7.94         527367          YES      109.60   
1            7.470           7.44         494055           NO      146.64   
2            7.515           7.44         547051           NO      147.88   
3            7.020           8.26         516279          YES      185.36   
4            7.070           8.26         531448           NO      176.48   

   Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0          223.840  Thriller              23            494       48000  
1          243.456     Drama              42            462       43200  
2         2022.400    Comedy              38            458       69400  
3          225.344     Drama              45            472       66800  
4          225.792     Drama              55            395       72400

We have 2 categorical data under: '3D_available' and 'Genre'

So, '3D_available' has Yes and No whereas 'Genre' has Thriller, Drama, Comedy, Action and we'll always take n-1 to create dummy variables

In [26]:
df = pd.get_dummies(df, columns = ["3D_available", "Genre"], drop_first = True, dtype = int)

In [27]:
df.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0         138.7               7.825                8.095            7.910   
1         152.4               7.505                7.650            7.440   
2         134.6               7.485                7.570            7.495   
3         119.3               6.895                7.035            6.920   
4         127.7               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views  Time_taken  Twitter_hastags  \
0            7.995           7.94         527367      109.60          223.840   
1            7.470           7.44         494055      146.64          243.456   
2            7.515           7.44         547051      147.88         2022.400   
3            7.020           8.26         516279      185.36          225.344   
4            7.070           8.26         531448      176.48          225.792   

   Avg_age_actors  Num_multiplex  Collection  3D_available_YES  Genre_Comedy  \
0              23            494       48000                 1             0   
1              42            462       43200                 0             0   
2              38            458       69400                 0             1   
3              45            472       66800                 1             0   
4              55            395       72400                 0             0   

   Genre_Drama  Genre_Thriller  
0            0               1  
1            1               0  
2            0               0  
3            1               0  
4            1               0

# Splitting data into X and Y variables

In [28]:
X = df.loc[:, df.columns != "Collection"]  #choosing all rows with : and all columns except Collection column

In [29]:
X.shape

(506, 19)

In [31]:
X.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0         138.7               7.825                8.095            7.910   
1         152.4               7.505                7.650            7.440   
2         134.6               7.485                7.570            7.495   
3         119.3               6.895                7.035            6.920   
4         127.7               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views  Time_taken  Twitter_hastags  \
0            7.995           7.94         527367      109.60          223.840   
1            7.470           7.44         494055      146.64          243.456   
2            7.515           7.44         547051      147.88         2022.400   
3            7.020           8.26         516279      185.36          225.344   
4            7.070           8.26         531448      176.48          225.792   

   Avg_age_actors  Num_multiplex  3D_available_YES  Genre_Comedy  Genre_Drama  \
0              23            494                 1             0            0   
1              42            462                 0             0            1   
2              38            458                 0             1            0   
3              45            472                 1             0            1   
4              55            395                 0             0            1   

   Genre_Thriller  
0               1  
1               0  
2               0  
3               0  
4               0

In [37]:
y = df["Collection"] #only choosing collection column

In [38]:
y.head()

0    48000
1    43200
2    69400
3    66800
4    72400
Name: Collection, dtype: int64

In [39]:
y.shape

(506,)

# Train and test data split

In [35]:
#only 20% of data is used as test data, and we'll train our model with 80% of data as our train data

In [36]:
from sklearn.model_selection import train_test_split

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [42]:
X_train.head() #just checking our X_train data, here we can notice that our index is changed against each row 

Marketing expense  Production expense  Multiplex coverage     Budget  \
220            27.1618               67.40               0.493  38612.805   
71             23.1752               76.62               0.587  33113.355   
240            22.2658               64.86               0.572  38312.835   
6              21.7658               70.74               0.476  33396.660   
417           538.8120               91.20               0.321  29463.720   

     Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
220         162.0               8.485                8.640            8.485   
71           91.0               7.280                7.400            7.290   
240         127.8               6.755                6.935            6.800   
6           140.1               7.065                7.265            7.150   
417         162.6               9.135                9.305            9.095   

     Producer_rating  Critic_rating  Trailer_views  Time_taken  \
220            8.670           8.52         480270      174.68   
71             7.455           8.16         491978      200.68   
240            6.840           8.68         470107      204.80   
6              7.400           8.96         459241      139.16   
417            9.165           6.96         302776      172.16   

     Twitter_hastags  Avg_age_actors  Num_multiplex  3D_available_YES  \
220          224.272              23            536                 0   
71           263.472              46            400                 0   
240          224.320              24            387                 1   
6            243.664              41            522                 1   
417          301.664              60            589                 1   

     Genre_Comedy  Genre_Drama  Genre_Thriller  
220             0            0               1  
71              0            0               0  
240             1            0               0  
6               0            0               1  
417             0            0               0

In [43]:
X_train.shape # X_train will be used to train our model

(404, 19)

In [44]:
X_test.shape # X_test will be used to check the performance of our model

(102, 19)

# Training regression tree

In [45]:
from sklearn import tree

In [46]:
regtree = tree.DecisionTreeRegressor(max_depth = 3) #never choose max depth greater than 5

In [47]:
regtree.fit (X_train, y_train)   #fitting the model

DecisionTreeRegressor(max_depth=3)

# Predict the values using trained model

In [49]:
y_train_pred = regtree.predict(X_train)

In [50]:
y_test_pred = regtree.predict(X_test)

In [51]:
y_test_pred

array([ 46187.35632184, 100000.        ,  46187.35632184,  23514.28571429,
        46187.35632184,  46187.35632184,  46187.35632184,  46187.35632184,
        46187.35632184,  46187.35632184,  23514.28571429,  23514.28571429,
        23514.28571429,  23514.28571429,  91860.86956522,  65183.33333333,
        46187.35632184,  65183.33333333,  46187.35632184,  46187.35632184,
        46187.35632184,  46187.35632184,  46187.35632184,  46187.35632184,
        46187.35632184, 100000.        ,  46187.35632184,  35137.77777778,
        91860.86956522,  35137.77777778,  23514.28571429,  23514.28571429,
        46187.35632184,  46187.35632184,  46187.35632184,  35137.77777778,
        23514.28571429, 100000.        ,  23514.28571429,  23514.28571429,
        46187.35632184,  46187.35632184,  46187.35632184,  23514.28571429,
        46187.35632184,  46187.35632184,  35137.77777778,  35137.77777778,
        35137.77777778,  46187.35632184,  35137.77777778,  35137.77777778,
        46187.35632184,  

# Evaluating model performance

In [52]:
from sklearn.metrics import mean_squared_error, r2_score

In [58]:
mean_squared_error(y_test, y_test_pred)

139740942.91784033

In [59]:
r2_score(y_train, y_train_pred)

0.8364651040556869

In [60]:
r2_score(y_test, y_test_pred)  #use test data to evaluate model performance

0.5659203159203307

In [61]:
# always r2 for train data set > r2 for test data set